plugins/pluginmanagers/lazy.nix

```
@@ -1,222 +1,204 @@
+      # plugin spec.
+      coerceToLazyPluginSpec =
+        plugin:
+        if lib.isDerivation plugin then
           {
-            name = "${lib.getName p}";
-            path = p;
+            dir = lib.mkDefault "${plugin}";
+            name = lib.mkDefault "${lib.getName plugin}";
           }
+        else if lib.isString plugin then
+          { __unkeyed = lib.mkDefault plugin; }
         else
-          {
-            name = "${lib.getName p.pkg}";
-            path = p.pkg;
-          };
-      processDependencies =
-        if plugin ? dependencies && plugin.dependencies != null then
-          builtins.concatMap processPlugin plugin.dependencies
-        else
-          [ ];
-    in
-    [ (mkEntryFromDrv plugin) ] ++ processDependencies;
-
-  processedPlugins = builtins.concatLists (builtins.map processPlugin lazyPlugins);
-  lazyPath = pkgs.linkFarm "lazy-plugins" processedPlugins;
-in
-{
-  options = {
-    plugins.lazy = {
-      enable = mkEnableOption "lazy.nvim";
+          plugin;
 
-      gitPackage = lib.mkPackageOption pkgs "git" {
-        nullable = true;
-      };
+      lazyPluginCoercableType = with types; either str package;
 
-      plugins =
+      lazyPluginType =
         with types;
-        let
-          pluginType = either package (submodule {
-            options = {
-              dir = helpers.mkNullOrOption str "A directory pointing to a local plugin";
-
-              pkg = mkOption {
-                type = package;
-                description = "Vim plugin to install";
-              };
+        types.coercedTo lazyPluginCoercableType coerceToLazyPluginSpec (
+          submodule (
+            { config, ... }:
+
+            let
+              cfg = config;
+            in
+            {
+              freeformType = attrsOf anything;
+              options = {
+                __unkeyed = mkNullOrOption str ''
+                  The "unkeyed" attribute is the plugin's short plugin url.
+                  It ill be expanded using `config.git.url_format`. It can
+                  also be a url or dir.
+                '';
+
+                dir = mkNullOrOption str "A directory pointing to a local plugin";
+
+                url = mkNullOrOption str "A custom git url where the plugin is hosted";
+
+                pkg = mkNullOrOption package "Vim plugin to install";
+
+                name = mkNullOrOption str "Name of the plugin to install";
+
+                dev = defaultNullOpts.mkBool false ''
+                  When true, a local plugin directory will be used instead.
+                  See config.dev
+                '';
+
+                lazy = defaultNullOpts.mkBool true ''
+                  When true, the plugin will only be loaded when needed.
+                  Lazy-loaded plugins are automatically loaded when their Lua modules are required,
+                  or when one of the lazy-loading handlers triggers
+                '';
```

```suggestion
                  When `true`, the plugin will only be loaded when needed.
                  
                  Lazy-loaded plugins are automatically loaded when their Lua modules are required,
                  or when one of the lazy-loading handlers triggers.
                '';
```

Do we want to explain which triggers these are ? I see there are multiple options so maybe a description at the top of the module would be good to explain how to use this?

---

Sorry, I wasn't very clear. I see that this is basically a core concept of the plugin and how you can control when things get lazy loaded through the various handlers we have options for. You can define a plugin level description with useful documentation like we do in [treesitter](https://github.com/nix-community/nixvim/blob/facf6b2d0c9e22d858956d1d458eac6baf155a08/plugins/by-name/treesitter/default.nix#L15) and some others. 

We might also benefit by saying a short form of this description which settings those are: 
```suggestion
                  When true, the plugin will only be loaded when needed.
                  Lazy-loaded plugins are automatically loaded when their Lua modules are required,
                  or when one of the lazy-loading handlers are triggered.
                  
                  You can define which triggers load this plugin using `plugins.lazy.plugins.<plugin>.[event|cmd|ft|keys]`.
                '';
```